include/ginkgo/core/multigrid/mapping.hpp

```diff
@@ -0,0 +1,158 @@
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/multigrid/multigrid_base.hpp>
+
+
+namespace gko {
+namespace multigrid {
+
+
+template <typename ValueType, typename IndexType>
+class Mapping : public EnableLinOp<Mapping<ValueType, IndexType>>,
+                public EnableCreateMethod<Mapping<ValueType, IndexType>>,
+                public ApplyAddable {
+    friend class EnableCreateMethod<Mapping>;
+    friend class EnablePolymorphicObject<Mapping, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    /**
+     * Returns the mapping indexes.
+     *
+     * @return the mapping indexes.
+     */
+    index_type *get_mapping_index() noexcept { return mapping_.get_data(); }
+
+    /**
+     * @copydoc Mapping::get_mapping_index()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const index_type *get_const_mapping_index() const noexcept
+    {
+        return mapping_.get_const_data();
+    }
+
+    /**
+     * Returns true as it is restrict operation.
+     *
+     * @return true as it is restrict operation.
+     */
+    bool is_restrict() const noexcept { return is_restrict_; }
+
+    /**
+     * Returns the number of elements explicitly stored in the mapping index.
+     *
+     * @return the number of elements explicitly stored in the mapping index
+     */
+    size_type get_num_stored_elements() const noexcept
+    {
+        return mapping_.get_num_elems();
+    }
+
+protected:
+    /**
+     * Creates an uninitialized Mapping arrays on the specified executor..
+     *
+     * @param exec  Executor associated to the LinOp
+     */
+    Mapping(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Mapping>(exec), mapping_{exec}, is_restrict_(true)
+    {}
+
+    template <typename IndicesArray>
+    Mapping(std::shared_ptr<const Executor> exec, const dim<2> &size,
+            IndicesArray &&mapping_indices, const bool &is_restrict = true)
+        : EnableLinOp<Mapping>(exec, size),
+          mapping_{exec, std::forward<IndicesArray>(mapping_indices)},
+          is_restrict_(is_restrict)
+    {
+        if (is_restrict_) {
+            GKO_ASSERT_EQ(size[1], mapping_.get_num_elems());
+        }
+        if (!is_restrict_) {
+            GKO_ASSERT_EQ(size[0], mapping_.get_num_elems());
+        }
+    }
+
+    void apply_impl(const LinOp *b, LinOp *x) const override
+    {
+        auto dense_x = as<matrix::Dense<ValueType>>(x);
+
+        // TODO: write dense fill kernel
+        auto zero = gko::initialize<matrix::Dense<ValueType>>(
+            {gko::zero<ValueType>()}, this->get_executor());
+        dense_x->scale(lend(zero));
```

Comment:
You can use the dense fill now.

Reply:
Yes. In fact, using scale is not always correct here if considering the initial vector of b is nan.
That's why I had duplicated work as you did :)